### PR TITLE
unit-testing

### PR DESCRIPTION
**Pull request by khasim-osv**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -71,27 +74,46 @@
...
     "ts-loader": "^6.2.1",
     "tslint": "^6.1.0",
     "tslint-immutable": "^6.0.1",
     "typescript": "^3.8.3",
     "webpack": "^4.42.0",
     "webpack-cli": "^3.3.11",
     "webpack-dev-server": "^3.10.3"
+  },
+  "jest": {
+    "automock": false,
```
Comment (manafay) [inline package.json:105]: setup the collectCoverage settings here....we can run and cover it once we merge it to master

File: src/app/common/datatable/Datatable.component.tsx [hunk 2]
```diff
@@ -102,7 +102,7 @@ const Datatable = (props: IProps) => {
         return onDefaultFilterMethod(filter, row, column);
       }}
       columns={columns}
-      defaultPageSize={state.pageSize}
+      defaultPageSize={3}
```
Comment (manafay) [inline src/app/common/datatable/Datatable.component.tsx:105]: why do we have 3 here?

File: src/app/common/datatable/Pagination.component.tsx [hunk 3]
```diff
@@ -69,38 +69,32 @@ export default class Pagination extends React.Component<IProps> {
...
-    );
-    for (
-      let i = 0;
-      i < Math.ceil(this.props.data.length / this.props.defaultPageSize);
-      i++
-    ) {
+    this.paging = [];
+    let len = this.props.data.length / this.props.defaultPageSize;
+    len % 1 === 0 ? (len = len - 1) : (len = Math.ceil(len));
+    for (let i = 0; i < len; i++) {
```
Comment (manafay) [inline src/app/common/datatable/Pagination.component.tsx:87]: no for loops please. use map or forEach. In very rare case we need then for..in could be used.

File: src/app/components/login/Login.redux.tsx [hunk 4]
```diff
@@ -10,7 +10,7 @@ export const mapStateToProps = (state: any) => {
 
 export const mapDispatchToProps = (dispatch: any) => {
   return {
-    onLogin: (userDetails: any) => dispatch(log_in(userDetails))
+    onLogin: (userDetails: any) => dispatch(login(userDetails))
```
Comment (manafay) [inline src/app/components/login/Login.redux.tsx:13]: why do we have any here? support with types.

File: src/app/components/login/Login.saga.js [hunk 5]
```diff
@@ -1,19 +1,26 @@
...
     res.data.success
-      ? yield put({ type: LOG_IN_SUCCEEDED, User: res.data })
-      : yield put({ type: LOG_IN_FAILED, User: res.data });
+      ? yield put({ type: LOGIN_SUCCEEDED, user: res.data })
+      : yield put({ type: LOGIN_FAILED, user: res.data });
   } catch (err) {
-    yield put({ type: LOG_IN_FAILED, User: { success: false, data: [] } });
+    yield put({
+      type: LOGIN_FAILED,
+      user: { success: false, data: {} }
```
Comment (manafay) [inline src/app/components/login/Login.saga.js:19]: you need not pass user object here...by default we will have default values in reducer...remove it.

File: src/app/components/master_data_management/translations/Translations.component.tsx [hunk 6]
```diff
@@ -20,6 +20,7 @@ export interface IProps {
   popupEditClose: Function;
   deleteTranslation: any;
   popupDeleteClose: Function;
+  translationDataChanged: Function;
```
Comment (manafay) [inline src/app/components/master_data_management/translations/Translations.component.tsx:23]: Again no types....and Function types also must be specific not just Function

File: src/app/components/master_data_management/translations/Translations.redux.tsx [hunk 7]
```diff
@@ -25,6 +26,8 @@ export const mapDispatchToProps = (dispatch: any) => {
     deleteTranslationData: (Id: any) => dispatch(delete_translation_data(Id)),
     popupEditClose: () => dispatch(edit_closed()),
     popupDeleteClose: () => dispatch(delete_closed()),
+    translationDataChanged: (data: Object) =>
```
Comment (manafay) [inline src/app/components/master_data_management/translations/Translations.redux.tsx:29]: this is not acceptable...Object, Function, Array.... Object of what, Function what it returns, Array of what values??

File: src/app/reducers.tsx [hunk 8]
```diff
@@ -14,7 +14,7 @@ import {
 } from "./components/master_data_management/modules/Modules.reducer";
 
 export const reducers = combineReducers({
-  User: login_reducer,
+  user: login_reducer,
   Translation: combineReducers({
     GetTranslations: getTranslationData_reducer,
```
Comment (manafay) [inline src/app/reducers.tsx:19]: why is there a need to have multiple reducers in translation?

File: src/app/components/master_data_management/translations/Translations.component.tsx [hunk 9]
```diff
@@ -111,8 +115,11 @@ const Translations = (props: IProps) => {
   ) {
     deleteModalToggle();
     setTimeout(() => {
+      props.translationDataChanged({
```
Comment (manafay) [inline src/app/components/master_data_management/translations/Translations.component.tsx:118]: No to setTimeout....think how you can achieve it.

File: src/app/components/master_data_management/translations/Translations.component.tsx [hunk 10]
```diff
@@ -99,8 +100,11 @@ const Translations = (props: IProps) => {
   ) {
     editModalToggle();
     setTimeout(() => {
```
Comment (manafay) [inline src/app/components/master_data_management/translations/Translations.component.tsx:102]: No to setTimeout....think how you can achieve it.

File: src/app/components/login/Login.redux.tsx [hunk 11]
```diff
@@ -10,7 +10,7 @@ export const mapStateToProps = (state: any) => {
 
 export const mapDispatchToProps = (dispatch: any) => {
```
Comment (manafay) [inline src/app/components/login/Login.redux.tsx:11]: we can even type dispatch..

File: src/app/components/login/__tests__/Saga.test.tsx [hunk 12]
```diff
@@ -0,0 +1,73 @@
+import { put, takeLatest } from "redux-saga/effects";
+import { loginUser, loginUserWatcher } from "../Login.saga";
+import { LoginActionTypes } from "../Login.actionTypes";
+
+jest.mock("axios");
+
+export interface IProps {
```
Comment (manafay) [inline src/app/components/login/__tests__/Saga.test.tsx:7]: unused part...remove it

File: src/app/components/master_data_management/translations/Translation.actionTypes.tsx [hunk 13]
```diff
@@ -0,0 +1,207 @@
+export enum TranslationActionTypes {
```
Comment (manafay) [inline src/app/components/master_data_management/translations/Translation.actionTypes.tsx:1]: Excellent work..I need this to be default approach in future

File: src/app/components/master_data_management/translations/Translations.action.tsx [hunk 14]
```diff
@@ -1,61 +1,62 @@
...
   return {
-    type: UPDATE_TRANSLATION_DATA,
-    payload: translationData
+    type: TranslationActionTypes.UPDATE_TRANSLATION_DATA,
+    payload: translationData,
   };
 }
 
-export function delete_translation_data(Id: any) {
+export function delete_translation_data(Id: string) {
```
Comment (manafay) [inline src/app/components/master_data_management/translations/Translations.action.tsx:25]: camelCase param names

File: src/app/components/master_data_management/translations/Translations.reducer.tsx [hunk 15]
```diff
@@ -1,122 +1,143 @@
...
+  TranslationReducerActions,
+  ITranslationData,
+  ITranslationDetails,
+  initialState,
+  getTranslationInitialState,
+} from "./Translation.actionTypes";
 
-export const getTranslationData_reducer = (state: any = {}, action: any) => {
+export const getTranslationData_reducer = (
+  state: ITranslationData = getTranslationInitialState,
```
Comment (manafay) [inline src/app/components/master_data_management/translations/Translations.reducer.tsx:11]: Needs to be followed at the rest of the places. Share this with the team

File: src/app/common/datatable/Datatable.component.tsx [hunk 16]
```diff
@@ -76,8 +95,8 @@ const Datatable = (props: IProps) => {
               rowInfo && rowInfo.index % 2 === 0 ? "#f3f5fd" : "#ffffff",
             fontFamily: "Avenir-Roman",
             fontSize: "16px",
-            color: "#384150"
-          }
+            color: "#384150",
```
Comment (manafay) [inline src/app/common/datatable/Datatable.component.tsx:98]: Brief the CSS design guy to have app level variable file for color codes & app level design elements(example responsive UI), we can import them here

Comment (manafay) [reply src/app/common/datatable/Datatable.component.tsx:98]: Even you can maintain it now

File: src/app/common/loaders/loader.component.tsx [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
+import "./loader.scss";
+
+export interface IProps {}
```
Comment (manafay) [inline src/app/common/loaders/loader.component.tsx:4]: empty prop type, why?

File: src/app/common/sidenavbar/SideNavbar.component.tsx [hunk 18]
```diff
@@ -17,12 +17,11 @@ export const SideNavbar = (props: IProps) => {
               <div>{props.title}</div>
             </li>
           )}
-          {props.List.map((obj) => {
+          {props.List.map(obj => {
```
Comment (manafay) [inline src/app/common/sidenavbar/SideNavbar.component.tsx:20]: can have return type of the array expected for predictability

File: src/app/components/master_data_management/translations/Translation.actionTypes.tsx (hunk 13 above)
Comment (khasim-osv) [reply src/app/components/master_data_management/translations/Translation.actionTypes.tsx:1]: Thanks Nafay, we are following the same

File: src/app/common/datatable/Datatable.component.tsx (hunk 16 above)
Comment (khasim-osv) [reply src/app/common/datatable/Datatable.component.tsx:114]: okay

**Pull request closed without merging**